10184: Add type annotations to twisted.web.template

mthuurne opened the pull request:
## Scope and purpose

The main goal is to provide type hints to projects using Twisted's template system or that are using just Stan.

Adding the type hints increased the amount of checking mypy can do on Twisted itself and this uncovered two bugs. One was fixed in #1590 + #1593, the other was pretty minor and is fixed in a dedicated commit in this PR.

Things to consider during the review:
(mostly copy-pasted from the commit comments)

Since a lot of functions consume or produce "Stan objects" or "objects can be flattened" and a lot of types can be flattened, I added a type alias for that: `Flattenable`. This may be useful for projects using Twisted as well, so I exported it as a public name from `twisted.web.template`.

The documentation for the `TagLoader` class used to state that it could load `IRenderable` providers, but the implementation actually accepts any flattenable object. The instance variable in which the to-be-loaded object is stored is even named `tag` while the `Tag` class does not implement `IRenderable`. Therefore I assume the docstring was incorrect and changed it to accept `Flattenable` instead. Note that `IRenderable` is a member of `Flattenable`, so the previously documented interface is supported by the annotation.

The `render` attribute of `Tag` will now raise `TypeError` when anything other than a string is assigned to it via `__call__`. Previously, the assignment was accepted but it would break during rendering.

The `request` argument to `twisted.web.template.renderElement()` can now be any `IRequest` implementer instead of only `twisted.web.server.Request`. The only reason it was limited to `server.Request` before was that the `site` attribute was inspected to decide whether to print a minimal or detailed error message. We now pick the minimal message for other request types, to be on the safe side (the detailed message might leak sensitive information).

For the `XMLFile` path, the annotation does not include types for which support is deprecated. I assume projects type checking their code do not want to use a feature that was deprecated 9 years ago.

I removed `@type` and `@rtype` docstring fields that provided no additional value over the type annotations. In quite a few cases, the documented types were incorrect, so this will increase the accuracy of the documentation.

## Contributor Checklist:

* [x] The associated ticket in Trac is here: https://twistedmatrix.com/trac/ticket/10184
* [x] I ran `tox -e lint` to format my patch to meet the [Twisted Coding Standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html)
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: [News files](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles))
* [x] The title of the PR starts with the associated Trac ticket number (without the `#` character).
* [x] I have updated the automated tests and checked that all checks for the PR are green.
* [x] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.
* [x] The merge commit will use the below format
  The first line is automatically generated by GitHub based on PR ID and branch name.
  The other lines generated by GitHub should be replaced.

```
Merge pull request #123 from twisted/4356-branch-name-with-trac-id

Author: mthuurne
Reviewer: wsanchez
Fixes: ticket:10184

Add type annotations to `twisted.web.template`.
```

mthuurne committed:
Do not hide failure in renderElement()'s finish handler

By returning the result, failures are preserved.
Thanks to @wiml for suggesting this solution.

mthuurne committed:
Fix renderElements()'s errback

The flatten() function returns a `Deferred`, so we can't call
`encode()` on it. We wouldn't need to anyway, since the flattened
data is already `bytes` (an UTF-8 encoded string).

This got broken in the Python 3 port, commit 55f72311 to be exact.
Since the failure was swallowed by the finish handler after all
output had already been produced, no-one noticed.

mthuurne committed:
Update Deferred.__await__() workaround to avoid mypy errors on use

While the original workaround succeeded in suppressing a mypy error
when `Deferred` is type-checked, the suppression also means that mypy
does not have correct type information on `Deferred.__await__()` and
as a result we get errors when we try to use it:

```py
from twisted.internet.defer import Deferred

async def f(d: Deferred[str]) -> None:
    s = await d  # error: Cannot determine type of '__await__'  [has-type]
    print(s)
```

mthuurne committed:
Make type annotations mandatory for twisted.web.template

Now that we have full mypy coverage, we should keep it that way.

mthuurne committed:
Raise TypeError when non-string renderer name is set on Tag

The `render` attribute of `Tag` will now raise `TypeError` when
anything other than a string is assigned to it via `__call__`.
Previously, the assignment was accepted but it would break during
rendering.

mthuurne committed:
Remove documentation of non-existing Expose.attributeName

The attribute name is hardcoded to "exposedThrough"; nothing was ever
assigned to the documented instance variable.

Since the `@renderer` decorator is the only user of the `Expose` class,
I don't think it's worthwhile to have a customizable attribute name,
so I chose to remove the `ivar` docstring field rather than implement
the feature.

mthuurne committed:
Simplify and fix docstring preservation by '@exposer'

The docstring preservation hack isn't necessary for pydoctor.
However, it is still good to have proper docstrings at runtime.
Because the value was assigned to `doc` instead of `__doc__`,
the docstring preservation was broken at runtime.

mthuurne committed:
Add type annotations for twisted.web.template

Since a lot of functions consume or produce "Stan objects" or "objects
that can be flattened" and a lot of types can be flattened, I added
a type alias for that: `Flattenable`. This may be useful for projects
using Twisted as well, so I exported it as a public name from
`twisted.web.template`.

For the `XMLFile` path, the annotation does not include types for
which support is deprecated. I assume projects type checking their
code do not want to use a feature that was deprecated 9 years ago.

The documentation for the `TagLoader` class used to state that it
could load `IRenderable` providers, but the implementation actually
accepts any flattenable object. The instance variable in which
the to-be-loaded object is stored is even named `tag` while the
`Tag` class does not implement `IRenderable`. Therefore I assume
the docstring was incorrect and changed it to accept `Flattenable`
instead. Note that `IRenderable` is a member of `Flattenable`, so
the previously documented interface is supported by the annotation.

I removed `@type` and `@rtype` docstring fields that provided no
additional value over the type annotations. In quite a few cases,
the documented types were incorrect, so this will increase the
accuracy of the documentation.

mthuurne committed:
Use attrs on 'slot' class

mthuurne committed:
Use attrs on Tag class

mthuurne committed:
Use attrs on CDATA, Comment and CharRef class

mthuurne committed:
Add type annotations to unit tests of twisted.web.template

mthuurne committed:
Improve type annotations for DeferredList and gatherResults()

Since `DeferredList` creates a new list from the given `Deferred`s,
there is no need to require an actual `list` to be passed: any
iterable is fine.

The previous annotations used `object` as the result type, but as
the type argument is contravariant, `Deferred[object]` was the only
valid specialization that could be passed.

With the new annotations, mypy can derive the result type from
the type of the argument passed to the `DeferredList` constructor
or to `gatherResults()`'s first argument.

A complication was the `fireOnOneCallback` flag, which selects
between two different result types. I don't think it is possible
to express this for a class in Python's type system, so I added
a workaround using overloaded function definitions.

mthuurne committed:
Raise ValueError when using a render method without a render factory

The previous implementation would try to use `None` as the render
factory, resulting in `AttributeError("'NoneType' object has no
attribute 'lookupRenderMethod'")`.

mthuurne committed:
Exclude 'raise NotImplementedError' from coverage

These are lines that we don't expect to be ever called.

mthuurne committed:
Link to all types mentioned in renderElement()'s docstring

This uncovers an ambiguity: there are 6 classes named `Request`
in Twisted and at least two of those would make sense here:
`twisted.web.http.Request` and `twisted.web.server.Request`.
When the type name is a link, pydoctor will warn about the
ambiguity.

The correct type is `twisted.web.server.Request`, since the error
handler accesses `Request.site`, which is unique to that class.
The docstring now refers to this class by its full name.

mthuurne committed:
Make renderElement() accept any IRequest

The only reason the request was limited to `twisted.web.server.Request`
was that the error handler used the `site` attribute to check whether
it should output a detailed backtrace or a minimal error message.

If a different `IRequest` implementer is passed, we pick the safe
option and output a minimal error message.

wsanchez commented on src/twisted/internet/defer.py:
```diff
@@ -959,7 +962,11 @@ def send(self, value: object = None) -> "Deferred[_DeferredResultT]":
     # type note: base class "Awaitable" defined the type as:
     #     Callable[[], Generator[Any, None, _DeferredResultT]]
     #     See: https://github.com/python/typeshed/issues/5125
```
Calling out https://github.com/python/typeshed/issues/5125 in this PR

wsanchez commented on src/twisted/internet/defer.py:
```diff
@@ -1159,10 +1166,43 @@ def __cmp__(self, other: object) -> int:
-_DeferredListResult = Tuple[bool, object]
+if TYPE_CHECKING:
 
+    # The result type is different depending on whether fireOnOneCallback
+    # is True or False.  The type system is not flexible enough to handle
+    # that in a class definition, so instead we pretend that DeferredList
+    # is a function that returns a Deferred.
```
OK.  Shouldn't there be an `else` before the class def? I guess the `type: ignore` deal with that?

mthuurne replied:
I picked `type: ignore` over `else` because this makes mypy check the implementation of `DeferredList`.

wsanchez replied:
Aha, cool. I didn't know that was even an option.

wsanchez commented on src/twisted/internet/defer.py:
```diff
@@ -1222,12 +1262,12 @@ def __init__(
         """
         self._deferredList = list(deferredList)
-        self.resultList: List[Optional[_DeferredListResult]] = [None] * len(
+        self.resultList: List[Optional[Tuple[bool, _DeferredResultT]]] = [None] * len(
```
A comment here explaining why `bool` is in here would be helpful, I think.

mthuurne replied:
I'll add a docstring for `resultList`.

wsanchez commented on src/twisted/web/_element.py:
```diff
@@ -19,15 +22,9 @@ class Expose:
-    tracked.
     """
 
-    def __init__(self, doc=None):
-        self.doc = doc
```
Is this a public class? That is, is this perhaps a compat issue?

mthuurne replied:
`Expose` is in a private module and not re-exported in any public module, so the class itself is private API.

wsanchez commented on src/twisted/web/_element.py:
```diff
@@ -58,13 +55,24 @@ def frob(self, a, b):
-            fObj.exposedThrough.append(self)
+            exposedThrough = getattr(fObj, "exposedThrough", [])
+            exposedThrough.append(self)
+            setattr(fObj, "exposedThrough", exposedThrough)
```
If this is to avoid a "no such attribute error" from Mypy, I wonder whether we prefer a `type: ignore` (which notes we are intentionally overriding the type system) is better. @glyph ^^

My thinking is that `type: ignore` invites occasional scrutiny in the future, and maybe that's worthwhile?

mthuurne replied:
I was going to argue that `getattr()` allows type checking of the attribute, but I see that I didn't actually annotate `exposedThrough`. I think that's the cleaner solution though: `type: ignore` will make the warning go away but will also prevent mypy from checking the code.

I'll add an annotation of `exposedThrough`. If it is later decided that `type: ignore` is preferred, that's fine too.

wsanchez commented on src/twisted/web/_element.py:
```diff
@@ -181,3 +181,7 @@ def render(self, request):
         return loader.load()
+
+
+if TYPE_CHECKING:
+    from twisted.web.template import Flattenable, Tag
```
Curious why not move this up so you don't need to put quotes around them in the file.

glyph replied:
Just my 2¢ here, I'd rather not see type annotation PRs ever introduce additional runtime coupling if they can possibly avoid it. When we adopt 3.7 we can drop all the quotes with `from __future__ import annotations`

glyph replied:
(In the meanwhile they seem like a small price to pay to avoid having to think hard about circular dependencies or import times)

mthuurne replied:
Circular dependencies are indeed a potential issue here: `template` imports `_element`, but it does so at the end of the module, so it would probably be fine but it would be brittle.

Therefore I'd rather leave the `TYPE_CHECKING` in place. ~I could move the statement to the top of the file, if you think that makes it easier to read.~ I have moved the import to the top of the module.

wsanchez commented on src/twisted/web/_flatten.py:
```diff
@@ -12,29 +12,72 @@
+T = TypeVar("T")
 
+FlattenableRec = Any
+"""
+For documentation purposes, read C{FlattenableRec} as L{Flattenable}.
+However, since mypy doesn't support recursive type definitions (yet?),
+we'll put Any in the actual definition.
+"""
```
Yeah I think JSON has this problem also.

wsanchez replied:
What's `Rec` stand for?

glyph replied:
Presumably `Recursive` ?

mthuurne replied:
> Yeah I think JSON has this problem also.

Every tree structure has this problem.

I once tried to do some tricky workaround and that crashed mypy. The crash was fixed quickly, but there is no solution for recursive types yet.

> What's `Rec` stand for?

> Presumably `Recursive` ?

Yes, recursive. I'll write it out in full to make it clearer.

wsanchez replied:
Cool.  I'm glad that you get some non-nested type checking here as a baseline.

wsanchez commented on src/twisted/web/_stan.py:
```diff
@@ -210,13 +214,13 @@ def _clone(self, obj, deep):
         """
         if hasattr(obj, "clone"):
-            return obj.clone(deep)
+            return obj.clone(deep)  # type: ignore[attr-defined, no-any-return]
```
We probably should `cast` instead of `no-any-return`, I think.

mthuurne replied:
Agreed.

wsanchez commented on src/twisted/web/_stan.py:
```diff
@@ -210,13 +214,13 @@ def _clone(self, obj, deep):
+            return obj.clone(deep)  # type: ignore[attr-defined, no-any-return]
         elif isinstance(obj, (list, tuple)):
-            return [self._clone(x, deep) for x in obj]
+            return [self._clone(x, deep) for x in obj]  # type: ignore[return-value]
```
`cast` here also, I think

mthuurne replied:
Hmm, strictly speaking we don't preserve the original type here: if a tuple is passed, we return a list instead. Is it acceptable to paper over this detail? Or would it be better to tone down the claim that `_clone()` can clone an arbitrary object and instead annotate it as taking and returning `Flattenable`?

I have a preference for the latter, since `_clone()` is made for tag contents and cannot actually clone arbitrary Python objects. For example if you'd tell it to make a deep clone of a JSON object, it would make a shallow copy of the dictionaries. I'll implement this option so it can be reviewed.

mthuurne replied:
Iterators are considered flattenable. Currently they are just copied, but they're not immutable. Should they be cloned into lists instead? We'd have to modify the original as well though, since we'd empty the iterator while cloning it, which might not be what the user expects.

Coroutines are also flattenable types, but I don't think we could clone those in any meaningful way. Should we just copy them as-is or raise an exception? Maybe raising an exception is the right solution for iterators too.

wsanchez replied:
Oof. Both sound problematic.  Since you can only drain an iterator once, if you copy it, one of the copies is going to get a different (empty) result when flattened, right?

wsanchez replied:
I vote for exceptions.

wsanchez commented on src/twisted/web/_stan.py:
```diff
@@ -288,54 +292,51 @@ def __repr__(self) -> str:
         return "CharRef(%d)" % (self.ordinal,)
+
+
+if TYPE_CHECKING:
+    from twisted.web.template import Flattenable
```
Why not import at the top?

mthuurne replied:
I think originally I didn't have the `TYPE_CHECKING` there and then importing at the bottom made sense to reduce the chance of cyclic import problems. I'll move it.

wsanchez commented on src/twisted/web/iweb.py:
```diff
@@ -825,3 +825,7 @@ def endpointForURI(uri):
 ]
+
+
+if TYPE_CHECKING:
+    from twisted.web.template import Flattenable, Tag
```
Top?

wsanchez commented on src/twisted/web/test/_util.py:
```diff
@@ -55,9 +62,9 @@ def assertFlattensImmediately(self, root, target):
         @rtype: L{bytes}
         """
-        return self.successResultOf(self.assertFlattensTo(root, target))
+        return self.successResultOf(self.assertFlattensTo(root, target))  # type: ignore[no-any-return]
```
`cast` instead

mthuurne replied:
I don't think I agree in this case. What we're working around here is not a limitation of mypy or the type system, but the fact that `successResultOf()` isn't annotated yet. I assume that will change at some point and having a `type: ignore` comment means that mypy will flag the suppression as a warning as soon as it can be removed. If we'd use a `cast`, we wouldn't be notified when the workaround becomes obsolete.

wsanchez replied:
Ah, ok

wsanchez replied:
An unnecessary `cast` will also cause complaint later, which is what I was angling for.  But I think you are right and both will.

wsanchez commented on src/twisted/web/template.py:
```diff
@@ -109,40 +126,43 @@ class _ToStan(handler.ContentHandler, handler.EntityResolver):
         Initialise the document.
         """
-        self.document = []
+        # Depending on our active context, the element type can be Tag, slot
+        # or str. Since mypy doesn't understand that context, it would be
+        # a pain to not use Any here.
+        self.document: List[Any] = []
```
"Leave as future exercise" seems totally reasonable in this PR

glyph replied:
I am not strongly recommending this, just mentioning it, but something that can sometimes be useful in this type of situation is adding a top-level type alias and commenting there, rather than at the call site; so like at module scope doing `TagSoupTerminal = Any` (or a better / more specific name if you can think of one) and then eventually replace it with `TagSoupTerminal = Union[Tag, slot, str]` (or similar) without changing the implementation logic

mthuurne replied:
> "Leave as future exercise" seems totally reasonable in this PR

I think this is a case where annotations alone aren't going to be sufficient; it will probably require some restructuring of the parser to make it more mypy-friendly.

> I am not strongly recommending this, just mentioning it, but something that can sometimes be useful in this type of situation is adding a top-level type alias and commenting there, rather than at the call site; so like at module scope doing `TagSoupTerminal = Any` (or a better / more specific name if you can think of one) and then eventually replace it with `TagSoupTerminal = Union[Tag, slot, str]` (or similar) without changing the implementation logic

That sounds like a good strategy in other places, but here it wouldn't save much, since there are only two annotations with `Any` and they're both in the block directly below this comment. If this were a public class, there might be some value in assigning a name for documentation purposes (*), but it's a private class instead.

(*) I think support for documented aliases should be added to pydoctor. That's useful for type aliases in particular; for other objects aliases are typically used to export an object under a different name and in that case the documentation of the original can be moved.

wsanchez replied:
Agreed with both ;-)

wsanchez commented on src/twisted/web/test/test_flatten.py:
```diff
@@ -237,13 +258,13 @@ def test_serializedAttributeWithTagWithAttribute(self):
         self.assertFlattensImmediately(Comment("foo bar"), b"<!--foo bar-->")
 
-    def test_commentEscaping(self):
+    def test_commentEscaping(self) -> Deferred[Any]:
```
`Deferred[object]` perhaps.

mthuurne replied:
`Deferred[object]` is actually not accepted by mypy. But in the error message, it tells me the actual type is `Deferred[List[bytes]]`, so I'll use that.

wsanchez replied:
I'm a little surprised about that, but ok

wsanchez commented on src/twisted/web/test/test_flatten.py:
```diff
@@ -442,15 +473,15 @@ class FlattenerErrorTests(SynchronousTestCase):
 
         @implementer(IRenderable)
-        class Renderable:
+        class Renderable:  # type: ignore[misc]
```
What's the `misc` error here?

mthuurne replied:
```
src/twisted/web/test/test_flatten.py:484:9: error: 'Renderable' is missing following 'IRenderable' interface members: lookupRenderMethod, render.  [misc]
```

wsanchez reviewed (approved):
Looks great, thanks for the work.
A few questions and nits, but nothing I'd block on.

mthuurne committed:
Add docstring for DeferredList.resultList

mthuurne committed:
Annotate exposedThrough local variable

mthuurne committed:
Rename FlattenableRec to FlattenableRecursive for more clarity

mthuurne committed:
Only claim that Tag._clone() can clone flattenables

The original code claimed to be able to clone arbitrary objects, but
that is not true in any meaningful way, since it is only able to make
deep copies of the types it knows.

mthuurne committed:
Move imports within a TYPE_CHECKING guard to the top of the module

Since the import is guarded, it cannot cause problems with circular
imports.

mthuurne committed:
Remove an unnecessary use of Any

mthuurne committed:
Merge branch 'trunk' into 10184-annotate-templates

mthuurne commented:
I think the only issue left is how to deal with `Tag._clone()`:
- do you agree with the change from arbitrary type to flattenable?
- should we raise an exception for flattenable types that are neither cloneable nor immutable, such as iterators and coroutines?

Regarding the latter, for new code I'd be in favor of rejecting input we cannot handle properly. But I'm worried that there is code out there that clones a tag and then only uses the clone, in that case we'd break a working program if we'd reject the cloning of iterators and coroutines.

wsanchez commented:
> I think the only issue left is how to deal with `Tag._clone()`:
> 
> * do you agree with the change from arbitrary type to flattenable?
> * should we raise an exception for flattenable types that are neither cloneable nor immutable, such as iterators and coroutines?

I think so.  And I think we can take some latitude here since it's private; we can fix it if it upsets something else in Twisted, but that seems a non-issue.

> Regarding the latter, for new code I'd be in favor of rejecting input we cannot handle properly. But I'm worried that there is code out there that clones a tag and then only uses the clone, in that case we'd break a working program if we'd reject the cloning of iterators and coroutines.

Hrm.  But that's *so* dangerous…  If not an exception, then maybe emitting a warning, so be replaced with an exception later, like a deprecation.

mthuurne commented:
> > I think the only issue left is how to deal with `Tag._clone()`:
> > 
> > * do you agree with the change from arbitrary type to flattenable?
> > * should we raise an exception for flattenable types that are neither cloneable nor immutable, such as iterators and coroutines?
> 
> I think so. And I think we can take some latitude here since it's private; we can fix it if it upsets something else in Twisted, but that seems a non-issue.

It's not as private as its name suggests, since it's called by the public `clone()` method to do a lot of the actual cloning.

> > Regarding the latter, for new code I'd be in favor of rejecting input we cannot handle properly. But I'm worried that there is code out there that clones a tag and then only uses the clone, in that case we'd break a working program if we'd reject the cloning of iterators and coroutines.
> 
> Hrm. But that's _so_ dangerous… If not an exception, then maybe emitting a warning, so be replaced with an exception later, like a deprecation.

I like the idea of a deprecation: it won't break applications but it will alert developers that they're doing something dangerous.

wsanchez committed:
Merge branch 'trunk' into 10184-annotate-templates

mthuurne committed:
Deprecate cloning of Tags containing generator or coroutine

mthuurne committed:
Make deprecation test pass CI

mthuurne commented:
The coverage regression is in `DNSDatagramProtocol.query()`, which should not be impacted at all by this PR. So I'm assuming that method is covered by a flaky tests.

wsanchez commented:
@mthuurne looks good, thanks!
